### PR TITLE
Improvements and Tests

### PR DESCRIPTION
**Pull request by martijn00**
## Connection with issue(s)

Connected to #4 

## To Do

- [x] Check the original issue to confirm it is fully satisfied
- [x] Add solution description to help guide reviewers
- [x] Add unit test to verify new or fixed behavior
- [x] If apply, add documentation to code properties and package readme

### CODE REVIEW

**Comment by caseycrogers** [comment]
Will wait on release! Just lmk when this is ready for merge and we're ready for launch

**Comment by caseycrogers** [comment]
Will wait on release! Just lmk when this is ready for merge and we're ready for launch

Though heads up I'm leaving for a ~6 day camping trip in 4 hours so it'll be awhile before I get back to this

**Comment by martijn00** [comment]
@caseycrogers Done! I'll open a new PR if i find anything but this should be good to merge and release.

File: lib/src/gutter_padding.dart
Comment (caseycrogers) [inline lib/src/gutter_padding.dart]: It feels really weird to me to pass Widgets in to configure padding sizes. I think this  feature (a gutter padding) would be VERY useful, but I think the API is hard to get right.

It's pretty tedious to write, but I think the best API might be:
`GutterPaddingLarge.only(<bools for each direction>)`
`GutterPaddingLarge.symmetric(<bools>)`
`GutterPaddingLarge.all()`
and repeated for each size Gutter. But I don't love this either, it's still a bit clunky and I wonder if it's that much better than just using the built in padding and edge inset system with `context.gutter` etc...

All in all I lean towards not merging this now until we come up with an API that feels clean. Happy to soundboard with you one what the API should look like, and if we do find something that feels good I don't mind doing the implementation myself if you don't feel like spinning up more PRs.

Can you snip this widget from the PR and I'll merge and ship?

Comment (martijn00) [reply lib/src/gutter_padding.dart]: I used a double before but that is not great either because the default should be to calculate the right gutter size when not hardcode to anything. This worked great for me but i agree it should rather not be a Widget itself. I'll remove it out for now and work in a new PR on it.

**Comment by martijn00** [comment]
@caseycrogers Done

File: lib/src/gutter_padding.dart
Comment (martijn00) [reply lib/src/gutter_padding.dart]: #7

File: lib/src/gutter_padding.dart
Comment (martijn00) [inline lib/src/gutter_padding.dart]: Using Gutter:
```dart
const GutterPadding.all(
    value: Gutter.medium(),
    child: ColoredBox(color: Colors.blue, child: Text('Child')),
),

const GutterPadding.only(
    left: Gutter.medium(),
    right: Gutter.small(),
    top: Gutter(size: 20),
    bottom: Gutter.medium(scaleFactor: 3),
    child: ColoredBox(color: Colors.blue, child: Text('Child')),
),
```

Using double? as parameter:

```dart
const GutterPadding.all(
    value: context.gutterSize(
      type: GutterType.medium,
    ),
    child: ColoredBox(color: Colors.blue, child: Text('Child')),
),

const GutterPadding.only(
    left: null,
    right: context.gutterSize(
      type: GutterType.small,
    ),
    top: context.gutterSize(
      type: GutterType.medium,
      size: 20
    ),
    bottom: context.gutterSize(
      type: GutterType.medium,
      scaleFactor: 3
    ),
    child: ColoredBox(color: Colors.blue, child: Text('Child')),
),
```

**Comment by martijn00** [comment]
@caseycrogers if you have some time when you are back can we look at getting this merged?

I'm also making some PRs to flutter adaptive scaffold itself to make some checks in here easier.

File: lib/src/gutter_padding.dart
Comment (caseycrogers) [reply lib/src/gutter_padding.dart]: I still am really not happy with `GutterPadding`, can we remove it for this PR?

I like the idea in theory, but:
A. Having a widget that takes in a child widget of a specific type and then extracts a value from it and throws away the widget feels really awkward
B. Alternatively, if it takes in a double, then what is the point of the widget anyway? Then it's just a `Padding` widget.

Right now what I do in my apps I think is more elegant than either of the two examples you show above. I think there may be some ways to come up with a cleaner Gutter specific API, but I don't think we've landed on it yet.
```dart
Padding(
  padding: EdgeInsets.only(
    right: context.gutterSmall, // <- `gutterSize` is unnecessary
    top: 20, // <- no need for `gutterSize`, if we're hardcoding the size anyway
    bottom: 3*context.gutter, // <- just do the math ourselves
  ),
);
```

The only thing `GutterPadding` saves us in the above example is avoiding the extra layer of nesting introduced by `EdgeInsets`, but really that's a problem with Flutter's core API more than it is a feature to be provided by this package.

File: lib/src/gutter_extensions.dart
Comment (caseycrogers) [inline lib/src/gutter_extensions.dart:40]: Lets remove this method and just put the logic in-line in the Gutter widgets and/or as a private method on Gutter. (basically the way it was before with `calculateGapSize`.

It really shouldn't be a publicly exposed part of the package as it is really kludgy to use and I don't see it providing any value over `context.gutterSmall` etc and then doing the math yourself.

File: lib/src/gutter_extensions.dart
Comment (caseycrogers) [inline lib/src/gutter_extensions.dart]: Sorry to really drag this out but lets get rid of this extension as well.

IMO it doesn't make any sense if you already have an extension for separating a list by an arbitrary widget (which most people do). It especially doesn't make sense given that we know we'll want to deprecate it in a release or two. It's also IIRC explicitly discouraged by the Flutter style guide to have methods that take in a subtype of Widget instead of just Widget as it damages composition.

File: lib/src/gutter.dart
Comment (caseycrogers) [inline lib/src/gutter.dart:13]: Can we make this constructor private and then have the public constructor be:
```dart
const Gutter({super.key, this.scaleFactor = 1})
      : type = GutterType.medium,
        size = null;
```

I have a really hard time imagining why a user of this package would want to manually specify the type and exposing `type` and `size` in a single constructor is really confusing because they two arguments are mutually exclusive.

Then we can add `Gutter.custom(this.size): ...` or something for the cases where someone using the API really does want to manually specify a size. (note that the above does not include scale factor as there's no point in exposing a scale factor if we're already manually specifying size and can just do the multiplication ourself)

File: lib/src/gutter.dart
Comment (caseycrogers) [inline lib/src/gutter.dart]: I'm really confused what this is. Lets remove it and you can explain it to me in a later issue/PR if you still feel like you want it in.

File: lib/src/gutter_type.dart
Comment (caseycrogers) [inline lib/src/gutter_type.dart]: I should've caught this earlier, but, what's the point of this GutterType? It feels like a buggier version of the `Expanded()` widget.

Lets remove it.

File: lib/src/gutter.dart
Comment (caseycrogers) [inline lib/src/gutter.dart:48]: I've thought about it a lot and I want to go back to the old API where there was a base `Gap` widget and then a bunch of specific `Gutter*` Widgets on top of that.
Alternate constructors are surfaced really poorly by IDE autocomplete UNTIL you type out `Gutter.`. This makes them much less discoverable and much more tedious to use.
Next, this gives us a clear way to have a custom `Gap` widget and a totally independent set of widgets for setting the default gap size. The current system in this PR is quite confused where if I want to manually set the size I create a default Gutter and then set an override size.

The counterpoints I see are:
1. It makes the implementation code sloppier
2. It puts two widgets in the tree per Gutter instead of 1

For 1, I don't think that's a super high priority relative to external dev ex. For 2, the perf cost should be relatively low and there is so much other low-hanging perf fruit in this package that doesn't come at the cost of dev ex that I'd much rather do first if perf is a concern (namely, ripping off the `Gap` package and creating the gaps via a render object instead of a `SizedBox`.


This is a pretty significant refactor undoing a good bit, some of which I already merged. As such, lets not make this change in this PR, it's quite sloppy but I'll make the change myself in a separate PR after merging this one.

**Review by caseycrogers — CHANGES_REQUESTED** [review]
I feel bad dragging this out so much but I don't think this PR is in a state I feel good about yet. My main requested changes are removing new features. Part of the problem is that the PR includes a refactor of the implementation details, a refactor of the external API, and several unrelated new features so it becomes hard to get out the door so I'm proposing more or less removing all the new features.

Once you remove the things I mentioned I'll merge this and then do some quick housecleaning (EDIT: actually some fairly heavy changes, namely rolling back to the `GutterTiny`/`GutterSmall`/etc system) on top of it and ship it out.

If you're burnt out on working on the PR given how much back and forth this has been let me know and I'll do all the changes I requested myself and merge and ship-I really appreciate your help and feel bad for taking up so much of your time.

File: lib/src/gutter_extensions.dart
Comment (martijn00) [reply lib/src/gutter_extensions.dart:40]: This is used in the tests and cannot be private. I would like to keep it in.

File: lib/src/gutter.dart
Comment (martijn00) [reply lib/src/gutter.dart:13]: I've now changed it to include `Gutter.from`

File: lib/src/gutter.dart
Comment (martijn00) [reply lib/src/gutter.dart:48]: I really don't think this is a good idea to revert back. This way is very common in the code bases i see. I've added classes for the sizes back in so you can use either way now.

I've thought about using a `RenderBox` as well but that was something i would like to do in a new PR.

**Comment by martijn00** [comment]
I've tried to accommodate your change requests. I really don't think rolling back to the GutterTiny etc and Gap is a good idea. Rather move forward and go to a RenderBox.

**Review by caseycrogers — APPROVED** [review]
There's a good bit of stuff I want to go through in this but I'm just going to merge it and do it on my end. I'll do my changes in a PR so you can see what I changed and what will eventually ship.

WRT to rolling back to Gap etc vs going forward to RenderBox, the former is the external APi and the latter is a perf centric implementation detail. IMO the two are totally independent-we should definitely switch the implementation to a RenderBox, it's just a matter of getting to it. I prefer the Gap/multiple widget API, but would love to hear why you prefer the single widget multiple constructors approach so we can consider migrating the external API to that

**Comment by martijn00** [comment]
I've made https://github.com/flutter/packages/pull/7300 here you can see the usage of multi constructors as well: https://github.com/flutter/packages/pull/7300/files#diff-8399b95990f0c1ed9192ac5d83957d04277e6d3b8b5add66bed1bb43b608f9e9R119

After https://github.com/flutter/packages/pull/7347 is merged I'll make a PR to update gutter to use it as well.

I'll also open some PR's to the adaptive package to add official sizes for the gap, hopefully that lands sometime soon.

It's ok if you prefer multiple widget, but i've made both ways possible right now, so why not leave both ways in?

**Pull request merged**
